README
NAME
    Net::Abuse::Utils - Routines useful for processing network abuse

VERSION
    version 0.23

SYNOPSIS
        use Net::Abuse::Utils qw( :all );
        print "IP Whois Contacts: ", join( ' ', get_ipwi_contacts($ip) ), "\n";
        print "Abuse.net Contacts: ", get_abusenet_contact($domain), "\n";

DESCRIPTION
    Net::Abuse::Utils provides serveral functions useful for determining
    information about an IP address including contact/reporting addresses,
    ASN/network info, reverse dns, and DNSBL listing status. Functions which
    take an IP accept either IPv6 or IPv4 IPs unless indicated otherwise.

CONFIGURATION
    There is a @RESOLVERS package variable you can use to specify name
    servers different than the systems nameservers for queries from this
    module. If you intend to use Google's nameservers here, please see This
    issue on GitHub for a note of caution
    <https://github.com/mikegrb/Net-Abuse-Utils/issues/9#issuecomment-243874
    35>.

FUNCTIONS
    The following functions are exportable from this module. You may import
    all of them into your namespace with the ":all" tag.

  get_asn_info ( IP )
    Returns a list containing (ASN, Network/Mask, CC code, RIR, modified
    date) for the network announcing "IP".

  get_peer_info ( IP )
    IPv4 Only. Returns an array of hash references containing (ASN,
    Network/Mask, CC code, RIR, modified date) for the peers of the network
    announcing "IP".

  get_as_description ( ASN )
    Returns the AS description for "ASN".

  get_as_company ( ASN )
    Similiar to "get_as_description" but attempts to clean it up some before
    returning it.

  get_soa_contact( IP )
    Returns the SOA contact email address for the reverse DNS /24 zone
    containing "IP".

  get_ipwi_contacts( IP )
    Returns a list of all email addresses found in whois information for
    "IP" with duplicates removed.

  get_rdns( IP )
    Returns the reverse PTR for "IP".

  get_dnsbl_listing( IP, DNSBL zone )
    IPv4 Only. Returns the listing text for "IP" for the designated DNSBL.
    "DNSBL zone" should be the zone used for looking up addresses in the
    blocking list.

  get_ip_country( IP )
    Returns the 2 letter country code for "IP".

  get_asn_country( ASN )
    Returns the 2 letter country code for "ASN".

  get_abusenet_contact ( domain )
    Returns the abuse.net listed contact email addresses for "domain".

  is_ip ( IP )
    Returns true if "IP" looks like an IP, false otherwise.

  get_domain ( IP )
    Takes a hostname and attempts to return the domain name.

  get_malware ( md5 )
    Takes a malware md5 hash and tests it against
    http://www.team-cymru.org/Services/MHR. Returns a HASHREF of last_seen
    and detection_rate.

DIAGNOSTICS
    Each subroutine will return undef if unsuccessful. In the furture,
    debugging output will be available.

CONFIGURATION AND ENVIRONMENT
    There are two commented out lines that can be uncommented to enable
    Memoize support. I haven't yet decided whether to include this option by
    default. It may be made available in the future via an import flag to
    use.

DEPENDENCIES
    This module makes use of the following modules:

    Net::IP, Net::DNS, Net::Whois::IP, and Email::Address

BUGS AND LIMITATIONS
    There are no known bugs in this module. Please report problems to
    Michael Greb (mgreb@linode.com)

    Patches are welcome.

ACKNOWLEDGEMENTS
    This module was inspired by Karsten M. Self's SpamTools shell scripts,
    available at http://linuxmafia.com/~karsten/.

    Thanks as well to my employer, Linode.com, for allowing me the time to
    work on this module.

    Rik Rose, Jon Honeycutt, Brandon Hale, TJ Fontaine, A. Pagaltzis, and
    Heidi Greb all provided invaluable input during the development of this
    module.

SEE ALSO
    For a detailed usage example, please see examples/ip-info.pl included in
    this module's distribution.

AUTHORS
    *   mikegrb <michael@thegrebs.com>

    *   Wes Young <github@barely3am.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Mike Greb.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Net::Abuse::Utils.

0.23  2013-10-24 19:08:48 EDT
    - [wesyoung] regex fix to get_peer_info where a date is not returned
    - minor build changes

0.22  2013-10-17 06:14:33 EDT
    - Pass prefix length to Net::IP::ip_reverse

0.21  2013-10-13 21:25:01 EDT
    - Return early when we don't have an AS Descriotion, closes GitHub #10

0.20  2013-09-13 23:51:24 EDT
    - @RESOLVERS package variable to use custom DNS resolvers instead of system
      resolvers
    - Declare required Perl version in dist metadata files

0.19  2013-09-05 18:21:39 EDT
    - Update docs, indicate which functions are IPv4 only
    - IPv6 test coverage
    - Other misc testing improvments

0.18  2013-09-03 21:17:51 EDT
    - [Andrew Hoying] IPv6 Support thanks to Andrew Hoying

0.17  2013-08-27 16:59:21 EDT
    - Offline tests

0.16  2013-08-25 23:15:13 EDT
    - Require ONLINE_TESTS ENV variable be set for online tests.

0.15  2013-07-30 06:16:32 EDT
    - Remove old Build.PL... ooops!

0.14  2013-07-29 01:02:00 EDT
    - Convert to Milla for authoring tool

0.13 2012-12-06
    - account for Net::DNS >= 0.69 converting SOA email record to address

0.12_01 2012-09-24
    - [wesyoung.me] added get_malware function based on: http://www.team-cymru.org/Services/MHR

0.12 2012-08-29
    - [wesyoung.me] test false failure bugfix
    - [wesyoung.me] get_asn_info and get_ipwi_contacts bug fixes
      closes RT #73203 and RT #70209
    - comment out some old DNSBLs and reorder output in ip-info.pl
    - add Wes Young to authors

0.11_01 2012-07-30
    - [wesyoung.me] added get_peer_info function
    - Use /usr/bin/env perl for example script shebangs

0.11  Sun Sep 14 2009
	- Moar Tests
	- Check asn is a number in get_asn_country

0.10  Sun Dec 28 2008
	- New get_domain function that converts host name to domain name
	- Memoize support though commented out by default, enable by 
	  uncommenting the following two lines in lib/Net/Abuse/Utils.pm:
		# use Memoize;
		# memoize('_return_rr');
	  A future version will likely allow Memoization via an export tag.

0.09  Sun Jun 15 2008
	- Fix META.yml issues that were missed in 0.08

0.08  Fri Jun 13 2008
	- update output of get_as_description to make ARIN IPs match other
	  RIRs, HANDLE AS Org 
	  eg: THEPLANET-AS ThePlanet.com Internet Services, Inc.

0.07  Fri Mar 30 2008
	- update regexs used by get_as_company to remove trailing 'AS Number'
	  as well as 'AS'
	- remove stripping of AS Handle from ARIN AS Desc now that we have
	  get_as_company
	- Module::Signature module signing stuff
	- Add license to Makefile.PL
	- Use newer ExtUtils::MakeMaker to generate the new META.yml format

0.06  Fri May 30 2008
	- Added get_as_company which functions similiarly to
	  get_as_description but attempts to clean up the string found before
	  returning it
	- Update AS description for 21844 in t/Net-Abuse-Utils.t
	- Remove relays.ordb.org from the DNSBL list in examples/ip-info.pl

0.05  Sun Mar 19 2006
	- Update Makefile.PL to specify Perl 5.6.1, missed this last rev
	- Fixed a bug in looking up ASN when the same block is announced by
	  multiple ASNs.
	- Fixed a bug where looking up AS Org for ASN not in Cymru database would
	  generate warnings rather then just return undef

0.04  Fri Mar 10 2006
	- Added public is_ip function to check that a string looks like an IP
	- Added tests using Test::Pod::Coverage and Test::Pod (if installed)
	- Prefix private functions with _ for Test::Pod::Coverage
	- Require Perl 5.6.1 rather then 5.8.7

0.03  Sun Mar  5 2006
	- 0.02 was uploaded with a broken test

0.02  Sun Mar  5 2006
	- Fixed a bug where AS Description would only return data
	  for ASs assigned by ARIN.
	- In get_ipwi_contacts check that whoisip_query returns a
	  hash ref rather then just checking that it isn't an array ref.
	- a few minor formatting cleanups

0.01  Sat Mar  4 2006
	- Initial Release


examples/ip-info.pl
#!/usr/bin/perl

use strict;
use warnings;

use Net::Abuse::Utils qw( :all );

my %dnsbl = (
		'Spamhaus'      => 'sbl-xbl.spamhause.org',
		'SpamCop'       => 'bl.spamcop.net',
		'Relays ORDB'   => 'relays.ordb.org',
		'Relays VISI'   => 'relays.vsi.com',
		'Composite BL'  => 'cbl.abuseat.org',
		'Dynablock BL'  => 'dnsbl.njabl.org',
		'DSBL Proxy'    => 'list.dsbl.org',
		'DSBL Multihop' => 'multihop.dsbl.org',
		'SORBS OR'      => 'dnsbl.sorbs.net',
		'SPEWS L1'      => 'l1.spews.dnsbl.sorbs.net',
		'SPEWS L2'      => 'l2.spews.dnsbl.sorbs.net',
		'Blitzed OPM'   => 'opm.blitzed.org',
		);

my $ip = shift;

my $rdns = get_rdns($ip) || '';
print "IP Info:\n";
print "\tIP:         $ip\n";
print "\tRDNS:       $rdns\n";
print "\tIP Country: ", get_ip_country($ip), "\n";

print "\nAS Info:\n";
if (my @asn = get_asn_info($ip) ) {
    my $asn_org = get_as_description($asn[0]) || '';
    print "\tASN:        $asn[0] - $asn[1]\n";
    print "\tAS  Org:    $asn_org\n";
    print "\tAS Country: ", get_asn_country($asn[0]), "\n";
}
else {
    print "\tUnknown ASN\n";
}

print "\nDNSBL Listings:\n";
foreach my $bl (keys %dnsbl) {
    my $txt = get_dnsbl_listing($ip, $dnsbl{$bl}) || 'not listed';
    print "\t$bl:\t$txt\n";
}

my $soa_contact = get_soa_contact($ip) || 'not found';
print "\nContact Addresses:\n";
print "\tPTR SOA Contact: $soa_contact\n";
print "\tIP Whois Contacts: ", join (' ', get_ipwi_contacts($ip) ), "\n";

if ($soa_contact =~ /\@(\S+)$/) {
    print "\tAbuse.net ($1): ", get_abusenet_contact($1), "\n";
}


lib/Net/Abuse/Utils.pm
package Net::Abuse::Utils;

use 5.008007;
use strict;
use warnings;

use Net::DNS;
use Net::Whois::IP qw(whoisip_query);
use Email::Address;

require Exporter;

our @ISA = qw(Exporter);

our %EXPORT_TAGS = ( 'all' => [ qw(
	get_asn_info get_as_description get_soa_contact get_ipwi_contacts
	get_rdns get_dnsbl_listing get_ip_country get_asn_country get_abusenet_contact
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our $VERSION = '0.03';
$VERSION = eval $VERSION;

sub reverse_ip {
    my $ip = shift;
    my @quads = split ('\.', $ip);
    return join('.', reverse(@quads));
} 

sub return_rr {
    my $lookup  = shift;
    my $rr_type = shift;
    my $concat  = shift;

    my @result;

    my $res = Net::DNS::Resolver->new;

    my $query = $res->query($lookup, $rr_type);
    if ($query) {
            foreach my $rr ($query->answer) {
                if ($rr->type eq $rr_type) { 
                    if    ($rr_type eq 'TXT') {
                        push @result, $rr->txtdata;
                    }
                    elsif ($rr_type eq 'SOA') {
                        push @result, $rr->rname;
                    }
                    elsif ($rr_type eq 'PTR') {
                        push @result, $rr->ptrdname;
                    }
                    last if !$concat;
                }
            }
            
            if ($concat && $concat == 2) {
                return @result;
            }
            else { 
                return join ' ', @result;
            }
    }
    
    return; 
}

sub return_unique {
    my $array_ref = shift;
    my %unique_elements;

    foreach my $element (@$array_ref) {
        $unique_elements{ $element }++;
    }

    return keys %unique_elements;
}

sub strip_whitespace {
    my $string = shift;
    
    return unless $string;
    
    for ($string) {
        s/^\s+//;
        s/\s+$//;
    }
    
    return $string;
}

sub get_ipwi_contacts {
    my $ip = shift;

    my @addresses;
    my %unique_addresses;

    my $response = whoisip_query($ip);

    # whoisip_query returns array ref if not found
    return unless ref($response) eq 'HASH';
    
    foreach my $field (keys %$response) {
        push @addresses, Email::Address->parse($response->{$field});
    }

    @addresses = map { $_->address } @addresses;

    return return_unique (\@addresses);
}

sub get_asn_info {
    my $ip = shift;

    my $lookup    = reverse_ip($ip) . '.origin.asn.cymru.com';
    my @origin_as = return_rr($lookup, 'TXT', 2) or return;
    
    # 23028 | 216.90.108.0/24 | US | arin | 1998-09-25
    # 701 1239 3549 3561 7132 | 216.90.108.0/24 | US | arin | 1998-09-25
 
    my $smallest_netmask = 0;
    my ($smallest_asn, %data_for_asn);

    # surely there is a better way to do this, at least the split
    # fields are stored so they don't have to be split again ;)
    for my $asn_info (@origin_as) {
        my @fields  = split /\|/, $asn_info;
        my @network = split '/', $fields[1];
        
        # we don't expect multiple asns but the spec says they may occur, so
        # just use the first
        (my $asn = $fields[0]) =~ /(d+)/; 
        
        $data_for_asn{$asn} = [ @fields ];
        
        if ($network[1] > $smallest_netmask) {
            $smallest_netmask = $network[1];
            $smallest_asn     = $asn;
        }
    }

    return map { strip_whitespace($_) } @{ $data_for_asn{$smallest_asn} };
}

sub get_as_description {
    my $asn = shift;
    
    my @ASdata = split('\|', return_rr("AS${asn}.asn.cymru.com", 'TXT'));
    
    # for arin we get HANDLE - AS Org
    if ($ASdata[2] eq ' arin ') {
        return strip_whitespace (( split (/ - /, $ASdata[4], 2) )[1]);
    }
    else {
        return strip_whitespace $ASdata[4];
    }
    
    return;
}

sub get_soa_contact {
    my $ip = shift;

    my $lookup = reverse_ip($ip) . '.in-addr.arpa';
    $lookup =~ s/^\d+\.//;

    if ( my $soa_contact = return_rr($lookup, 'SOA') ) {
        $soa_contact =~ s/\./@/;
        return $soa_contact;
    }
    
    return;
}

sub get_rdns {
    my $ip = shift;

    return return_rr( reverse_ip($ip) . '.in-addr.arpa', 'PTR');
}

sub get_dnsbl_listing {
    my $ip    = shift;
    my $dnsbl = shift;

    my $lookup = join '.', reverse_ip( $ip ), $dnsbl;
    
    return return_rr($lookup, 'TXT', 1);
}

sub get_ip_country {
     my $ip = shift;
     
     return (get_asn_info($ip))[2];
}

sub get_asn_country {
    my $asn   = shift;
    my $as_cc = ( split (/\|/,return_rr("AS${asn}.asn.cymru.com", 'TXT') ) )[1];
    if ($as_cc) {
        return strip_whitespace ($as_cc);
    }
    return;
}

sub get_abusenet_contact {
    my $domain = shift;
    
    return return_rr("$domain.contacts.abuse.net", 'TXT', 1)
}

1;
__END__

=head1 NAME

Net::Abuse::Utils - Routines useful for processing network abuse


=head1 VERSION

This documentation refers to Net::Abuse::Utils version 0.03.


=head1 SYNOPSIS

    use Net::Abuse::Utils qw( :all );
    print "IP Whois Contacts: ", join( ' ', get_ipwi_contacts($ip) ), "\n";
    print "Abuse.net Contacts for $domain: ", get_abusenet_contact($domain), "\n";

=head1 DESCRIPTION

Net::Abuse::Utils provides serveral functions useful for determining
information about an IP address including contact/reporting addresses,
ASN/network info, reverse dns, and DNSBL listing status.


=head1 FUNCTIONS

The following functions are exportable from this module.  You may import all
of them into your namespace with the C<:all> tag.

=over 5

=item get_asn_info ( IP )

Returns a list containing (ASN, Network/Mask, CC code, RIR, modified date)
for the network announcing C<IP>.

=item get_as_description ( ASN )

Returns the AS description for C<ASN>. 

=item get_soa_contact( IP )

Returns the SOA contact email address for the reverse DNS /24
zone containing C<IP>.

=item get_ipwi_contacts( IP )

Returns a list of all email addresses found in whois information
for C<IP> with duplicates removed.

=item get_rdns( IP )

Returns the reverse PTR for C<IP>.

=item get_dnsbl_listing( IP, DNSBL zone )

Returns the listing text for C<IP> for the designated DNSBL.  C<DNSBL zone>
should be the zone used for looking up addresses in the blocking list.

=item get_ip_country( IP )

Returns the 2 letter country code for C<IP>.

=item get_asn_country( ASN )

Returns the 2 letter country code for C<ASN>.

=item get_abusenet_contact ( domain )

Returns the abuse.net listed contact email addresses for C<domain>.

=back

=head1 DIAGNOSTICS

Each subroutine will return undef if unsuccessful.  In the furture,
debugging output will be available.

=head1 CONFIGURATION AND ENVIRONMENT

No configuration mechanism is currently implemented.

=head1 DEPENDENCIES

This module makes use of the following modules:

Net::DNS, Net::Whois::IP, Email::Address

=head1 BUGS AND LIMITATIONS

There are no known bugs in this module.  Please report problems to
Michael Greb (mgreb@linode.com)

Patches are welcome.

=head1 ACKNOWLEDGEMENTS

This module was inspired by Karsten M. Self's SpamTools shell scripts, 
available at http://linuxmafia.com/~karsten/.

Thanks as well to my employer, Linode.com, for allowing me the time to work
on this module.

Rik Rose, Jon Honeycutt, Brandon Hale, TJ Fontaine, and A. Pagaltzis all
provided invaluable input during the development of this module.

=head1 AUTHOR

Michael Greb (mgreb@linode.com)

=head1 SEE ALSO

For a detailed usage example, please see examples/ip-info.pl included in
this module's distribution.

=head1 LICENCE AND COPYRIGHT

Copyright (c)  2006 Michael Greb (mgreb@linode.com). All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut